server/src/MapChunk.ts

```diff
@@ -67,13 +68,60 @@ export class PlayerChunkDB extends BaseEntity implements PlayerChunk {
 	@JoinColumn({ name: 'hash' })
 	data!: ChunkData
 
+	/**
+	 * Saves a chunk
+	 * @param map_chunk chunk
+	 */
 	static async store(map_chunk: PlayerChunk) {
 		// TODO if PlayerChunk exists, and holds last reference to old hash, delete ChunkData at old hash
-		await ChunkDataDB.upsert(map_chunk.data, {
-			conflictPaths: ChunkDataDB.primaryCols,
-			skipUpdateIfNoValuesChanged: true,
+
+		// create object to store all data needed for chunk data
+		const chunkData = {
+			data: map_chunk.data.data,
+			hash: map_chunk.data.hash,
+			version: map_chunk.data.version,
+			chunkX: map_chunk.chunk_x,
+			chunkZ: map_chunk.chunk_z,
+			timestamp: map_chunk.ts,
+			world: map_chunk.world,
+		}
+
+		// object for all player data minus timestamp because where can't take it
+		// added later when actually saving
+		const playerData = {
+			world: map_chunk.world,
+			chunkX: map_chunk.chunk_x,
+			chunkZ: map_chunk.chunk_z,
+			uuid: map_chunk.uuid,
+		}
+
+		prisma.playerChunk.upsert({
+			where: {
+				world_chunkX_chunkZ_uuid: playerData,
+			},
+			create: {
+				...playerData,
+				timestamp: map_chunk.ts,
+				chunkData: {
+					connectOrCreate: {
+						where: {
+							hash: map_chunk.data.hash,
+						},
+						create: chunkData,
+					},
+				},
+			},
+			update: {
+				...playerData,
+				timestamp: map_chunk.ts,
+				chunkData: {
+					upsert: {
+						create: chunkData,
+						update: chunkData,
```

Comment:
We should only set the timestamp if this ChunkData doesn't already exist, as that's the most accurate time when the chunk last changed. It should be the time this ChunkData was first created; actually that would also be a good column name.

Reply:
should be fixed in 92f0417